ot/mpz-ot-core/src/ferret/spcot/sender.rs

```diff
@@ -0,0 +1,262 @@
+            state: state::Extension {
+                delta,
+                unchecked_vs: Vec::default(),
+                vs_length: Vec::default(),
+                cot_counter: 0,
+                exec_counter: 0,
+                extended: false,
+                prg: Prg::from_seed(seed),
+                hasher: blake3::Hasher::new(),
+            },
+        }
+    }
+}
+
+impl Sender<state::Extension> {
+    /// Performs the SPCOT extension.
+    ///
+    /// See Step 1-5 in Figure 6.
+    ///
+    /// # Arguments
+    ///
+    /// * `h` - The depth of the GGM tree.
+    /// * `qs`- The blocks received by calling the COT functionality.
+    pub fn extend(
+        &mut self,
+        h: usize,
+        qs: &[Block],
+        mask: MaskBits,
+    ) -> Result<ExtendFromSender, SenderError> {
+        if self.state.extended {
+            return Err(SenderError::InvalidState(
+                "extension is not allowed".to_string(),
+            ));
+        }
+
+        if qs.len() != h {
+            return Err(SenderError::InvalidLength(
+                "the length of q should be h".to_string(),
+            ));
+        }
+
+        let MaskBits { bs } = mask;
+
+        if bs.len() != h {
+            return Err(SenderError::InvalidLength(
+                "the length of b should be h".to_string(),
+            ));
+        }
+
+        // Updates hasher.
+        self.state.hasher.update(&bs.to_bytes());
+
+        // Step 3-4, Figure 6.
+
+        // Generates a GGM tree with depth h and seed s.
+        let s = self.state.prg.random_block();
+        let ggm_tree = GgmTree::new(h);
+        let mut k0 = vec![Block::ZERO; h];
+        let mut k1 = vec![Block::ZERO; h];
+        let mut tree = vec![Block::ZERO; 1 << h];
+        ggm_tree.gen(s, &mut tree, &mut k0, &mut k1);
+
+        // Stores the tree, i.e., the possible output of sender.
+        self.state.unchecked_vs.extend_from_slice(&tree);
+
+        // Stores the length of this extension.
+        self.state.vs_length.push(1 << h);
+
+        // Computes the sum of the leaves and delta.
+        let sum = tree.iter().fold(self.state.delta, |acc, &x| acc ^ x);
+
+        // Computes M0 and M1.
+        let mut ms: Vec<[Block; 2]> = qs
+            .iter()
+            .zip(bs)
+            .map(|(&q, b)| {
+                if b {
+                    [q ^ self.state.delta, q]
+                } else {
+                    [q, q ^ self.state.delta]
+                }
+            })
+            .collect();
+
+        ms.iter_mut()
+            .enumerate()
+            .map(|(i, blks)| {
+                let tweak: Block = bytemuck::cast([i, self.state.exec_counter]);
+                let tweaks = [tweak, tweak];
+                FIXED_KEY_AES.tccr_many(&tweaks, blks);
+                blks
+            })
+            .zip(k0.iter().zip(k1.iter()))
+            .for_each(|([m0, m1], (k0, k1))| {
+                *m0 ^= *k0;
+                *m1 ^= *k1;
+            });
```

Comment:
I find this a bit hard to follow with the multiple closures, perhaps a simple loop?

```suggestion
        // Computes M0 and M1.
        let mut ms: Vec<[Block; 2]> = Vec::with_capacity(qs.len());
        for (((i, &q), b), (k0, k1)) in qs.iter().enumerate().zip(bs).zip(k0.into_iter().zip(k1)) {
            let mut m = if b {
                [q ^ self.state.delta, q]
            } else {
                [q, q ^ self.state.delta]
            };

            let tweak: Block = bytemuck::cast([i, self.state.exec_counter]);
            FIXED_KEY_AES.tccr_many(&[tweak, tweak], &mut m);

            m[0] ^= k0;
            m[1] ^= k1;

            ms.push(m);
        }
```

Reply:
Well, I used iter twice before to make the code more readable.
For functionality, all these ways are workable.

Reply:
Sorry for being picky :grinning: The multiple closures iterates over `ms` 3 times I think (not sure if the compiler optimizes it away)

Reply:
I think you are right on this point. Although I think the TCCR hash will dominate the performance, but I think it is always good to make possible optimizations.